dialog/dialog/relational-algebra.md

```diff
@@ -0,0 +1,415 @@
+
+For example:
+
+```
+rgb = [
+    (name: "Red"),
+    (name: "Green"),
+    (name: "Blue")
+]
+
+cmy = [
+    (name: "Cyan"),
+    (name: "Magenta"),
+    (name: "Yellow")
+]
+
+roygbiv = [
+    (name: "Red"),
+    (name: "Orange"),
+    (name: "Yellow"),
+    (name: "Green"),
+    (name: "Blue"),
+    (name: "Indigo"),
+    (name: "Violet")
+]
+
+union(rgb, cmy, roygbiv) => [
+    (name: "Red"),
+    (name: "Green"),
+    (name: "Blue"),
+    (name: "Cyan"),
+    (name: "Magenta"),
+    (name: "Yellow"),
+    (name: "Orange"),
+    (name: "Indigo"),
+    (name: "Violet")
+]
+```
+
+For simplicity, implementations MAY implement union as a binary operator, instead simulating n-ary unions through the composition of unions.
+
+For example, using the above relations:
+
+```
+union(rgb, cmy, roygbiv) <=> union(rgb, union(cmy, roygbiv))
+                         <=> union(union(rgb, cmy), roygbiv)
+                         <=> ...
+```
+
+## 3.5 Difference
+
+Difference is a binary operation which performs a set difference over its input relations. 
+
+All of the input relations must be union-compatible, by sharing the same set of attribute names.
+
+For example:
+
+```
+roygbiv = [
+    (name: "Red"),
+    (name: "Orange"),
+    (name: "Yellow"),
+    (name: "Green"),
+    (name: "Blue"),
+    (name: "Indigo"),
+    (name: "Violet")
+]
+
+rgb = [
+    (name: "Red"),
+    (name: "Green"),
+    (name: "Blue")
+]
+
+difference(roygbiv, rgb) => [
+    (name: "Orange"),
+    (name: "Yellow"),
+    (name: "Green"),
+    (name: "Indigo"),
+    (name: "Violet")
+]
+```
+
+## 3.6 Cartesian Product
+
+A Cartesian product is a binary operation which computes all possible pairs between two input relations. The relations MUST have disjoint attributes, and the tuples in the resulting set are flattened.
+
+For example:
+
+```
+ranks = [
+    (rank: "ace"),
+    (rank: "king"),
+    (rank: "queen"),
+    (rank: "jack")
+]
+
+suits = [
+    (suit: "spades"),
+    (suit: "hearts"),
+    (suit: "diamonds"),
+    (suit: "clubs")
+]
+
+cartesian_product(ranks, suits) => [
+    (rank: "ace", suit: "spades"),
+    (rank: "ace", suit: "hearts"),
+    (rank: "ace", suit: "diamonds"),
+    (rank: "ace", suit: "clubs"),
+    (rank: "king", suit: "spades"),
+    (rank: "king", suit: "hearts"),
+    (rank: "king", suit: "diamonds"),
+    (rank: "king", suit: "clubs"),
+    (rank: "queen", suit: "spades"),
+    (rank: "queen", suit: "hearts"),
+    (rank: "queen", suit: "diamonds"),
+    (rank: "queen", suit: "clubs"),
+    (rank: "jack", suit: "spades"),
+    (rank: "jack", suit: "hearts"),
+    (rank: "jack", suit: "diamonds"),
+    (rank: "jack", suit: "clubs")
+]
+
+```
+
+## 3.7 Natural Join
+
+A natural join is a binary operator which computes the set of all combinations of the tuples in its input relations, such that the values of any attributes common to both relations are equal.
```

Comment:
This is a really great question to ask, and what you're proposing is actually a path I'd started to go down in my implementation, until I realized why it isn't as elegant as it first appears.

First, consider a rule for computing the 1-hop neighbors in a graph:
```
oneHop(src: Src, dst: Dst) :- edge(src: Src, dst: Via), edge(src: Via, dst: Dst).
```

Here, what we want is a join between `edge` and `edge`, on the attributes `(dst, src)`, with the result containing `src` from the first relation, and `dst` from the second relation, but with `src` and `dst` appearing in both input relations. This means that it's no longer as simple as returning the flattened combination of all attributes, but that in addition to specifying to which attributes in each input relation we're joining against, we also need to specify a projection + rename against both input relations. The full query plan ends up looking something like this:

```
  edge                                                                                          edge
   |                                                                                             |
   v                                                                                             v
project([src, dst]) -> theta_join(on: [left.dst = right.src], into: [left.src, right.dst]) <- project([src, dst])
```

Under the more traditional natural join approach, we end up with this query plan:

![image](https://user-images.githubusercontent.com/285821/199586083-09431ad0-4672-410f-bf3a-d478b517706e.png)

If you squint, it becomes clear that both of these query plans are actually equivalent, except that in the first we've moved a ton of complexity into the `theta_join`, rather than building everything up out of smaller composable pieces!

Also — it's a separate topic, but I'm not actually convinced that we want to support `theta_join` in the first place, because it doesn't work as nicely with the datastructures we're using for indexes, and I think it's probably simpler to just represent a `theta_join` as the composition of `select` and `natural_join`. This also frees up implementations to apply optimization passes that decompose and reorder the `select` operations so that they're applied as early as possible in the query, to speed up and reduce the memory overhead of complex joins.

Does this make sense? This is a great question, with a very subtle answer, so I think it's important to make sure we're on the same page here, because it's something I've run into myself, too :)

Reply:
> both of these query plans are actually equivalent, except that in the first we've moved a ton of complexity into the `theta_join`

Yeah... Right. That does make sense.

---

I still have a lingering feel here that the relational algebra level should already have abstracted away names, as they're mostly useful for users (and debugging?), but maybe not machines. In an implementation, shouldn't rho practically be a no-op?